editor/src/clj/editor/handler.clj

```diff
@@ -19,197 +19,264 @@
-            menu-item (assoc menu-item :command dynamic-id)]
-        (-> acc
-            (update-in [:ids id :dynamics] (fnil conj #{}) dynamic-id)
-            (register-handler dynamic-id dynamic-id context-definition (:fns handler))
-            (as-> x (reduce #(register-menu %1 dynamic-id %2 [menu-item]) x locations)))))
-    state
-    dynamic-handlers))
-
-(defn- unregister-dynamics [state id]
-  (reduce
-    (fn [acc dynamic-id]
-      (-> acc
-          (unregister-menu dynamic-id)
-          (unregister-handler dynamic-id)))
-    (util/dissoc-in state [:ids id :dynamics])
-    (get-in state [:ids id :dynamics])))
-
-(defonce state-atom (atom {}))
-
-(s/def ::menu-item
-  (s/keys :opt-un [::id ::icon ::style ::label ::command ::children ::graphic-fn ::user-data ::check]))
-
-(s/def ::menu (s/coll-of ::menu-item))
-(s/def ::id keyword?)
-(s/def ::icon string?)
-(s/def ::style (s/coll-of string?))
-(s/def ::label (s/or :string string? :separator #{:separator}))
+      (util/dissoc-in acc [location registration]))
+    location->registration->menu
+    locations))
+
+(defn- unregister [{:keys [registrations] :as state} registration]
+  (if-let [{:keys [command+contexts locations]} (registrations registration)]
+    (-> state
+        (assoc :registrations (dissoc registrations registration))
+        (update :handlers remove-handlers command+contexts registration)
+        (update :menus remove-menus locations registration))
+    state))
+
+(def ^:private synthetic-command-str "synthetic-command")
+
+(defn synthetic-command? [x]
+  (= (namespace x) synthetic-command-str))
+
+(defn- register [state registration menus handlers]
+  (let [handlers (mapv (fn [h]
+                         (if (contains? h :command)
+                           h
+                           (assoc h :command (keyword synthetic-command-str (str synthetic-command-str (RT/nextID))))))
+                       handlers)
+        menus (reduce
+                (fn [acc [location menu-item]]
+                  (update acc location (fnil conj []) menu-item))
+                menus
+                (e/mapcat
+                  (fn [{:keys [command locations]}]
+                    (let [menu-item {:command command}]
+                      (e/map #(coll/pair % menu-item) locations)))
+                  handlers))]
+    (-> state
+        (unregister registration)
+        (update :menus add-menus menus registration)
+        (update :handlers add-handlers handlers registration)
+        (update :registrations assoc registration {:locations (vec (keys menus))
+                                                   :command+contexts (vec (e/mapcat
+                                                                            (fn [{:keys [command contexts]}]
+                                                                              (e/map #(coll/pair command %) contexts))
+                                                                            handlers))}))))
```

Comment:
I refactored handler registration so it has more approachable semantics w.r.t. registration identifier, that were previously scoped differently in different places. Now there is a single `registration` id that is used to register both menus and handlers.